Add PaalmanPings to Polaris Total Scattering

DanielMurphy22 opened the pull request:
**Description of work.**

Adds PaalmanPings correction to total scattering workflow. This currently only works for POLARIS and sample shape cylinder (and slab). 

**To test:**

Ask me for the calculate-pdf example directory.
You will want to add these 2 lines to the polaris_config_example.yaml file:
```
empty_can_subtraction_method: 'PaalmanPings'
paalman_pings_events_per_point: 1
```

Launch MantidWorkbench and run the following script:
```python
from isis_powder import polaris, SampleDetails
import time

start_time = time.time()
config_file_path = r"/home/danielmurphy/Downloads/polaris-calculate-pdf/polaris_config_example.yaml"
polaris = polaris.Polaris(config_file=config_file_path, user_name="test", mode="PDF")

'''Slab'''
# sample_details = SampleDetails(height=4.0, width=4.0, thickness=4.0, angle=0.0, center=[0, 0, 0], shape='slab')
# sample_details.set_material(chemical_formula='Si')
# sample_details.set_container(front_thick=4.0, back_thick=4.0, chemical_formula='V')
# polaris.set_sample_details(sample=sample_details)
'''Cylinder'''
sample_details = SampleDetails(height=4.0, radius=0.2985, center=[0, 0, 0], shape='cylinder')
sample_details.set_material(chemical_formula='Si')
sample_details.set_container(radius=0.3175, chemical_formula='V')

polaris.create_vanadium(first_cycle_run_no="98532", multiple_scattering=False)
polaris.focus(run_number="98533", input_mode='Summed', sample_empty="98531")
polaris.create_total_scattering_pdf(run_number="98533", merge_banks=False)
```
Try the Slab and Cylinder shape options

Fixes #33030

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DanielMurphy22 committed:
Allow PaalmanPings for POLARIS and shape cylinder only

Save applied PaalmanPings to file
Take EventsPerPoints input
Also allow loading in polaris_algs from archive
by removing .nxs file extension from Load() commands
Explicitly use Minus algm in ApplyPaalmanPings
Create isis_powder container_details
Allow simple material setting in container/sample details

DannyHindson commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ApplyPaalmanPingsCorrection.py:
```diff
@@ -377,12 +378,17 @@ def _three_factor_corrections_approximation(self, sample_workspace, container_wo
         acsc = factor_workspaces['acsc']
         assc = factor_workspaces['assc']
-        return (sample_workspace - container_workspace * (acsc / acc)) / assc
+        subtrahend = container_workspace * (acsc / acc)
```
Subtrahend - that's a good word (and a new one to me)!

DannyHindson commented on scripts/Diffraction/isis_powder/polaris_routines/polaris_algs.py:
```diff
@@ -35,7 +35,7 @@ def _get_run_numbers_for_key(current_mode_run_numbers, key):
     mapping_dict = get_cal_mapping_dict(run_number_string, inst_settings.cal_mapping_path)
     if inst_settings.mode is None:
-        ws = mantid.Load('POLARIS'+run_number_string+'.nxs')
+        ws = mantid.Load('POLARIS'+run_number_string)
```
What's the reason for removing the .nxs prefix here?

DanielMurphy22 replied:
To allow loading from the archive "POLARIS98533.nxs" does not exist but searching "POLARIS98533" will find "POLARIS0098533.nxs" on the archive. We should check that this doesn't stop loading locally on Windows.

DannyHindson commented on scripts/Diffraction/isis_powder/routines/focus.py:
```diff
@@ -48,14 +51,43 @@ def _focus_one_ws(input_workspace, run_number, instrument, perform_vanadium_norm
-    if summed_empty is not None:
+
+    # Correct for absorption / multiple scattering if required
+    if paalman_pings and summed_empty and absorb and container_details:
```
I wonder if it might be better if we had an overall "absorb" flag that is a boolean and then a second parameter called something like "absorb_method" that is a string and takes values like "simple" and "paalman_pings" (with the default value being "simple"). That way it would be more obvious that setting absorb=false turns off all absorption corrections. What do you think?

DannyHindson commented on scripts/Diffraction/isis_powder/polaris.py:
```diff
@@ -72,13 +75,21 @@ def create_total_scattering_pdf(self, **kwargs):
+    def set_container_details(self, **kwargs):
+        self._set_environment_details("container", **kwargs)
+
+    def _set_environment_details(self, environment_type: str = "sample", **kwargs):
```
Based on the way you've set this up I wonder if it makes sense to rename the sample_details class to something general like shape_details and to use it for holding a shape and a material regardless of whether the object is the sample or a container.
I appreciate this is a rather different suggestion to what I was talking about earlier where I was wondering about having some kind of super class that could hold the details of both a sample and a container. I guess that was motivated partly by the idea of having some validation logic that checked that the sample and container shapes are consistent. But that probably ought to be checked in the PPMonteCarloAbsorption algorithm rather than you replicating that validation logic here.

DannyHindson commented on scripts/Diffraction/isis_powder/routines/focus.py:
```diff
@@ -48,14 +51,43 @@ def _focus_one_ws(input_workspace, run_number, instrument, perform_vanadium_norm
+        else:
+            corrections = mantid.PaalmanPingsMonteCarloAbsorption(
+                InputWorkspace=input_workspace,
+                Shape=sample_details._shape_type.capitalize(),
```
I think we should discuss this point - but I'll pose the question here at least:
In other parts of the code that sample_details class is used to prepare the inputs to a call to the SetSample algorithm. An absorption correction algorithm is then called soon after that.
I wonder if it would make things more consistent if you called SetSample here and then used the Shape=Preset option in the PaalmanPingsMonteCarloAbsorption algorithm

DanielMurphy22 committed:
Requested Changes

Merge Container into sampledetails
Add absorb method enum option for yaml file

DannyHindson commented on scripts/Diffraction/isis_powder/routines/focus.py:
```diff
@@ -66,23 +63,24 @@ def _focus_one_ws(input_workspace, run_number, instrument, perform_vanadium_norm
-                ContainerChemicalFormula=container_details.material_object.chemical_formula,
-                EventsPerPoint=events_per_point
-            )
+            mantid.SetSample(input_workspace, Geometry={'Shape': sample_details._shape_type.capitalize(),
```
Please can you see if you can add a method to the sample_details class called something like `get_sample_geometry` that generates this json string? I think that will make it easier to support the slab shape as well as the cylinder
This function will essentially be the same as `generate_sample_geometry` in common.py but I think it should be part of the sample_details class. As I mentioned in our call I think `generate_sample_geometry` should be removed once you've done this

DanielMurphy22 committed:
Add Slab container to ISISPowder SampleDetails

move generate_sample_geometry() and material() to sample_details
create generate_container_geometry() and material()

DanielMurphy22 commented:
I **think** I've implemented the Slab shape with appropriate container details, but running this script fails with a MonteCarloAbsorption error generting points within the shape:

```
from mantid.simpleapi import *
import numpy as np
from isis_powder import polaris, SampleDetails
import time

start_time = time.time()
config_file_path = r"/home/danielmurphy/Downloads/polaris-calculate-pdf/polaris_config_example.yaml"
polaris = polaris.Polaris(config_file=config_file_path, user_name="test", mode="PDF")

sample_details = SampleDetails(height=4.0, width=2.0, thickness=4.0, angle=90.0, center=[0, 0, 0], shape='slab')
sample_details.set_material(chemical_formula='Si')
sample_details.set_container(front_thick=1.0, back_thick=2.0, chemical_formula='V')
polaris.set_sample_details(sample=sample_details)

polaris.create_vanadium(first_cycle_run_no="98532", multiple_scattering=False)
polaris.focus(run_number="98533", input_mode='Summed')
polaris.create_total_scattering_pdf(run_number="98533", merge_banks=False)
  
print('\n\n')
print("Script run time", time.time() - start_time)
```

DanielMurphy22 commented on scripts/Diffraction/isis_powder/routines/focus.py:
```diff
@@ -48,22 +48,47 @@ def _focus_one_ws(input_workspace, run_number, instrument, perform_vanadium_norm
+            input_workspace = mantid.ApplyPaalmanPingsCorrection(SampleWorkspace=input_workspace,
+                                                                 CorrectionsWorkspace=corrections,
+                                                                 CanWorkspace=summed_empty)
+            mantid.SaveNexus(input_workspace, os.path.join(save_directory, expected_filename))
```
This saved file of the applied paalmanpings workspace has no knowledge of the sample/container shape/material so may be reloaded ignoring any changed input values. So loading a saved file should probably only happen if the inputs to SetSample() are identical

DannyHindson commented:
I've seen that kind of MCAbs error before when the sample and container shapes are both hollow eg donut shaped. Not sure why that's happening here though.

Does it still work OK on the cylinder sample shape + container? Perhaps an ideal thing to use the shape viewing on to check the slab shapes are as expected?

The changes you've done look good and are what I had in mind. I guess it might be nice to hang the functions that generate the json strings (eg generate_sample_geometry) off the _Slab and _Cylinder classes rather than having an if statement. You could do something like:
```
def generate_sample_geometry(self):
   return self._shape.generate_sample_geometry()
```

DanielMurphy22 committed:
Allow SetSample to accept int, long or double

Also removed the constraint on isis_powder.SampleDetails
for the slab angle to be greater than 0

DanielMurphy22 committed:
Move center from SampleDetails to _shape

create _shapes for HollowCylinder and FlatPlateHolder containers
and invoke generate_sample/container_geometry from the _shape

DanielMurphy22 committed:
Save PalmanPingsCorrections workspace and add force_recalculate switch

DanielMurphy22 committed:
Make angle parameter optional for SetSample FlatePlate

DanielMurphy22 committed:
Add SetSample and ISISPowder SetSample tests

DanielMurphy22 committed:
Merge branch 'main' into PalmaanPings_TotScat_33030

DanielMurphy22 committed:
Docs and release notes

DanielMurphy22 committed:
Move test_generate_sample_material/geometry to sampledetailstest

DanielMurphy22 commented:
Some additional thoughts: 
- Should the PaalmanPings lines I've added to focus.py really be in absorption_corrections.py ???
- Should the options for absorption method be "PaalmanPings" and "Mayers" (rather than "Simple"??)

DannyHindson commented:
> Some additional thoughts:
> * Should the PaalmanPings lines I've added to focus.py really be in absorption_corrections.py ???

Good question. I think ideally yes but looking at how the existing absorption correction code in absorb_corrections.py is called, it seems to be a two hop process. focus.py calls _apply_absorb_corrections on the instrument object and then the individual instruments (Pearl, Polaris, HRPD, GEM) call the functions in absorb_corrections.py. So achieving this might involve adding a function to the instrument as well
The extra level on the instrument seems to exist to allow both real sample runs and V runs to be corrected via a call to instrument._apply_absorb_corrections. Most of the _apply_absorb_corrections functions have an if statement that checks if the run is a V run and if yes automatically sets up the V shape\material.
This leads onto another question - the unit calibrate.py is used to focus V runs. It has a function in it called create_van which is very similar to focus. I think this should perhaps also have the same PP changes in it that you've done to focus.py
Let's discuss

> * Should the options for absorption method be "PaalmanPings" and "Mayers" (rather than "Simple"??)

Yes good idea

DannyHindson reviewed (changes requested):
I just built the code and ran the script:
- I'm getting an error about events_per_point_string in focus.py which is stopping the PP process completing. It looks like this variable isn't hooked up to the config yaml file setting:
![image](https://user-images.githubusercontent.com/56295817/166643677-87f61da9-fa32-4a8e-8d51-729b4a64ccd2.png)
- We need a test that runs a focus with the PP option enabled so that the previous point is caught by the automated tests. This could either go in the unit tests (these are currently quite sparse in terms of calls to focus but doesn't mean we should avoid that) or in a system test. The system tests are instrument specific so if you added a test there, it probably makes sense to add it to POLARIS
- when I run without having called sample_details.set_container the process fails with an error about NoneType. This should probably give a more helpful error message

DanielMurphy22 committed:
Requested Changes including adding Polaris Focus PaalmanPings systest

DannyHindson commented on scripts/Diffraction/isis_powder/routines/focus.py:
```diff
@@ -48,22 +48,37 @@ def _focus_one_ws(input_workspace, run_number, instrument, perform_vanadium_norm
-    # Correct for absorption / multiple scattering if required
     if absorb:
-        input_workspace = instrument._apply_absorb_corrections(run_details=run_details, ws_to_correct=input_workspace)
+        if summed_empty:
```
Couple of things about this if statement:

1) I think the condition for running PP should instead be `if run_details.sample_empty:`
This relates to the conversation we were having with Helen about the two types of "empty" run. If summed_empty contains the empty instrument run (ie no sample and no can) then we shouldn't be doing the PP approach. This potentially complicates the if\else statements over the next ~20 lines...

2) the rework of the logic over the next ~20 lines means that the route through the code for people not running the PP approach will change a bit. Most of it should be functionally equivalent - although the order of the cropping\absorption correction steps have swapped - but even so it introduces a bit of risk that we may break something. Could we do something a bit more like the following instead? As well as keeping the non-PP code closer to the original, it has a bit less duplication in it I think:
```
if absorb and PP:
  if run_details.sample_empty:
    instrument._apply_paalmanpings_absorb_and_subtract_empty
  else:
    error
else
  if summed_empty:
    common.subtract_summed_runs
  crop_raw_to_expected_tof_range
  if absorb:
    instrument._apply_absorb_corrections
```

DannyHindson reviewed (changes requested):
I've added a couple of comments about the new if\else statement in focus.py. If this is all a bit late in the day, then it's not especially important to put this into the release. It would be nice to round this off today\Monday though even if it just goes into main after the release is branched off it so you can clear the decks while it's all fresh in our minds.
Also one other detail level thing: what was the reason for the change to ApplyPaalmanPings? It was something about running this where one of the input workspace parameters was the same as the output but I forget exactly. Might be worth adding a release note for this.

DanielMurphy22 committed:
Requested Changes

add ISIS_PowderPolaris systeest FocusMayers FocusPaalmanPings systest
Rename absorb_method Simple to Mayers
Only implement PaalmanPings for Polaris
Reorganise logic in focus._focus_one_ws to maintain current simple workflow
Update error message when running PaalmanPings without sample empty
Add release note for ApplyPaalmanPingsCorrection bugfix

DannyHindson commented on scripts/Diffraction/isis_powder/abstract_inst.py:
```diff
@@ -67,7 +67,8 @@ def _focus(self,
                do_absorb_corrections,
-               sample_details=None):
+               sample_details=None,
+               absorb_method=None):
```
Not sure you need this absorb_method parameter now that you've removed it from the onward call to focus

DannyHindson commented on Testing/SystemTests/tests/framework/ISIS_PowderPolarisTest.py:
```diff
@@ -121,6 +121,92 @@ def cleanup(self):
+        self.tolerance = 1e-6
+        mantid.SaveNexus(self.focus_results, "/home/danielmurphy/Desktop/ISIS_Powder-POLARIS98533_FocusMayers.nxs")
+        return self.focus_results.name(), "ISIS_Powder-POLARIS98533_FocusMayers.nxs"
+        pass
```
Don't think this pass statement is needed?

DannyHindson commented on Testing/SystemTests/tests/framework/ISIS_PowderPolarisTest.py:
```diff
@@ -121,6 +121,92 @@ def cleanup(self):
+        self.tolerance_is_rel_err = True
+        self.tolerance = 1e-6
+        return self.focus_results.name(), "ISIS_Powder-POLARIS98533_FocusPaalmanPings.nxs"
+        pass
```
Don't think this pass statement is needed?

DannyHindson commented on scripts/Diffraction/isis_powder/routines/focus.py:
```diff
@@ -48,22 +48,38 @@ def _focus_one_ws(input_workspace, run_number, instrument, perform_vanadium_norm
+    try:
+        absorb_method = instrument._inst_settings.absorb_method
+    except AttributeError:
+        absorb_method = 'Mayers'
```
If you reinstated the `absorb_method` parameter on all the instruments would it remove the need for this? That might be nicer than doing this here.
The NotImplementedError inside abstract_inst.py would effectively turn off the PP feature on all the other instruments which is fine

DannyHindson commented on Testing/SystemTests/tests/framework/ISIS_PowderPolarisTest.py:
```diff
@@ -121,6 +121,92 @@ def cleanup(self):
+            self.assertEqual(ws.sample().getMaterial().name(), 'Si')
+        self.tolerance_is_rel_err = True
+        self.tolerance = 1e-6
+        mantid.SaveNexus(self.focus_results, "/home/danielmurphy/Desktop/ISIS_Powder-POLARIS98533_FocusMayers.nxs")
```
This line needs removing

DannyHindson commented on Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ApplyPaalmanPingsCorrection.py:
```diff
@@ -377,12 +378,17 @@ def _three_factor_corrections_approximation(self, sample_workspace, container_wo
         assc = factor_workspaces['assc']
-        return (sample_workspace - container_workspace * (acsc / acc)) / assc
+        subtrahend = container_workspace * (acsc / acc)
+        difference = Minus(sample_workspace, subtrahend)
```
I just noticed while running a focus with PP enabled that you end up with workspaces with the following names left behind in the ADS:
- subtrahend
- difference

Is it possible to clean these up. Or perhaps there's a way of stopping them being saved into the ADS in the first place?
I guess minuend would also get left behind (see _two_factor function bit lower down)
I think it's fine that the "corrections" workspace group that the CalculatePaalmanPings algorithm creates is left behind but the ones I mentioned above feels more like internal working that should be cleaned up

DannyHindson reviewed (changes requested):
Just had another look through. I've added a few more minor comments.
Not sure if Jenkins is very well because I can access the logs for the jobs that are currently running? I think one of my comments suggests that one of the system tests will fail if\when the jobs do complete
I'm going to stop for the day shortly but we can follow up on Monday. As I thought earlier I think we should probably not put this into the release but let's aim to get rounded off on Monday anyway

sf1919 commented:
The Release Notes for v6.5 have now been created. Please can you move your release note over to the correct folder before Monday 16th May?

DannyHindson commented on scripts/Diffraction/isis_powder/routines/absorb_corrections.py:
```diff
@@ -112,3 +114,47 @@ def _do_cylinder_absorb_corrections(ws_to_correct, multiple_scattering, is_vanad
+        logger.warning(f'Loading pre-calculated corrections file found at {expected_path} \n'
+                       f' To recalculate with new sample + container values,'
+                       f' set "force_recalculate_paalman_pings = True" in the config file.')
+        corrections = mantid.LoadNexus(expected_path)
```
I get an error when running a PP calculation where the corrections file already exists:

![image](https://user-images.githubusercontent.com/56295817/167398067-a6af723f-b156-44b6-ae87-8915327e8664.png)
The error comes from the ApplyPaalmanPingsCorrection call but I think it originates from the fact that this call to LoadNexus returns a tuple rather than a workspace group. This seems to be a feature of the C++ code for LoadNexus where it sets up multiple output workspace properties if the .nxs file is a group. The tuple has 5 things in it called OutputWorkspace, OutputWorkspace_1, OutputWorkspace_2 etc and the first one is the workspace group. So you probably want to set corrections equal to corrections.OutputWorkspace
Since the code path using the "cache" will be quite a common one I think it's worth having a test that goes through this bit of code as well. Fine to just run the calculation twice in the test

DannyHindson commented on Testing/SystemTests/tests/framework/ISIS_PowderPolarisTest.py:
```diff
@@ -121,6 +121,92 @@ def cleanup(self):
+        pass
+
+    def cleanup(self):
+        mantid.mtd.clear()
```
I think you also need to delete the PP corrections .nxs file that is output. I just tried and it gets deleted by one of the other tests in this suite but if this test was run on its own the .nxs file would be left behind. I suggest you add the following lines in cleanup here that are present in the other tests:
```
        try:
            _try_delete(output_dir)
            _try_delete(spline_path)
        finally:
            mantid.mtd.clear()
            config['datasearch.directories'] = self.existing_config
```
The final line in here sets the data search config back to its original value and this is worth including so other system tests aren't affected by the code in this unit that sets this to some special POLARIS value

DannyHindson commented on Testing/SystemTests/tests/framework/ISIS_PowderPolarisTest.py:
```diff
@@ -121,6 +121,92 @@ def cleanup(self):
+        pass
+
+    def cleanup(self):
+        mantid.mtd.clear()
```
There's no PP corrections .nxs file to worry about here but you should align the clean function here with what is done in the other tests so any other files get cleaned up and the data search config back to its original value

DannyHindson reviewed (changes requested):
Just done a bit more testing on this while I've got the branch built. These fixes may be slightly more than can be squeezed in quickly today before we get onto release activities but see how you get on
Re my comments on Friday about not leaving certain workspaces behind in the ADS - I think if you add StoreInADS=False to the parameters list this will do the trick

DanielMurphy22 committed:
Improve Polaris systemtests

add test which runs loads the cached paalmanpngs correction file
cleanup files and workspaces from new tests
allow absorb_method as an optional param map attribute on all powder isntruments

DanielMurphy22 committed:
Merge branch 'main' into PalmaanPings_TotScat_33030

DanielMurphy22 committed:
Move release notes to 6.5

DanielMurphy22 committed:
Pass ApplyPaalmanPingsCorrection test

By storing returned workspaces from factor functions
in the ADS but hidden to allow the workspace passed to AddSampleLog
to have a name

DanielMurphy22 commented:
The [SystemTests.ISIS_PowderPolarisTest.FocusTestAbsorptionMayers](https://builds.mantidproject.org/job/pull_requests-rhel7/50042/testReport/junit/SystemTests/ISIS_PowderPolarisTest/FocusTestAbsorptionMayers/) seems to be passing locally.

DanielMurphy22 committed:
Increase tolerance on MayersAbsorption test to pass on Rhel7

DannyHindson reviewed (changes requested):
Thanks for the changes. This is looking really good. Apologies for spotting\raising the sample_empty_scale thing at this late stage. Hope it's not too problematic.

Couple of other general things:

- please can you add a bit of detail to ISISPowder-Polaris-v1.rst about the new switches (absorb_method, paalman_pings_events_per_point, force_recalculate_paalman_pings)? There's a section on the existing do_absorb_corrections setting in here
- the cached PP file is quite big (it's almost ~1GB for me on run 98533). Any idea why it's so large? It's a workspace group containing 4 workspaces so that's partly the answer but still not sure why each workspace is ~250MB. The nexus files for run 98533 is only 24MB. Maybe it contains the instrument definition?

DanielMurphy22 committed:
Do not cache corrections file

Update Docs with new switches absorb_method, paalman_pings_events_per_point
Set sample_empty scale factor to 1 when using PaalmanPings
Remove "workspace" from ADS

DannyHindson commented:
The latest code has failed the ISIS_PowderPolarisTest.FocusTestAbsorptionPaalmanPings system test on two of the Jenkins jobs. I think the problem is that the code in run_focus_absorption sets the sample_empty_scale to 0.5 and one of the code changes you've put in means that this is (correctly) ignored when the PP approach is turned on. I think you need to update the ref file for the PP test and I suggest you also only set the scale factor if PP=false. Developers sometimes look at these system tests as an example of how to run the algorithms so best the test follows the correct approach rather than relying on the setting being ignored.

DannyHindson commented on scripts/Diffraction/isis_powder/abstract_inst.py:
```diff
@@ -67,7 +67,9 @@ def _focus(self,
                do_absorb_corrections,
-               sample_details=None):
+               sample_details=None,
+               absorb_method="Mayers",
```
The HRPD instrument uses an algorithm called HRPDSlabCanAbsorption to calculate its absorption corrections so the Mayers method isn't perhaps used widely enough to have it as a default here. I think everything should work OK if you just use `None` as the default value here.
Also this has got me thinking - perhaps the absorb_method should actually be called empty_container_subtraction_method with values PaalmanPings and Simple. We have a few absorption correction algorithms using different types of calculation (eg MayersSampleCorrection and MonteCarloAbsorption are both Monte Carlo calculations, HRPDSlabCanAbsorption uses numerical integration and there are others that just uses pre-calculated lookup tables) and in theory they could all be used within a PaalmanPings approach or a simple empty can subtraction approach.

DannyHindson commented on scripts/Diffraction/isis_powder/routines/focus.py:
```diff
@@ -14,20 +14,24 @@
-def _focus_one_ws(input_workspace, run_number, instrument, perform_vanadium_norm, absorb, sample_details,
-                  vanadium_path):
+def _focus_one_ws(input_workspace, run_number, instrument, perform_vanadium_norm, absorb,
+                  sample_details, vanadium_path, absorb_method="Mayers", paalman_pings_events_per_point=None):
```
One of my other comments talks about changing the values in the absorb_method parameter (and perhaps the name of the parameter itself) so this comment may be overtaken by that...
If we do keep the absorb_method parameter as is, I don't think you need the default "Mayers" values in all 4 places in this unit. It's probably sufficient just to have the default on the outer focus function

DanielMurphy22 committed:
Merge branch 'main' into PalmaanPings_TotScat_33030

DanielMurphy22 committed:
Update test file FocusPaalmanPings

Also Removed unneccessary default arguments

DanielMurphy22 committed:
Rename absorb_method to empty_can_subtraction_method in code and docs

DannyHindson commented on docs/source/techniques/ISISPowder-Polaris-v1.rst:
```diff
@@ -400,6 +402,20 @@ Example Input:
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Sets the absorption method used in the :ref:`focus_polaris_isis-powder-diffraction-ref` method. This parameter
+is optional and can be set to either ``"Simple"`` (default) or ``"PaalmanPings"``. In ``"PaalmanPings"`` mode
+the :ref:`paalman_pings_events_per_point_isis-powder-diffraction-ref` parameter can be utilised.
```
Perhaps worth mentioning here that the PaalmanPings option results in the PaalmanPingsMonteCarloAbsorption algorithm being run followed by ApplyPaalmanPingCorrection

DannyHindson commented on scripts/Diffraction/isis_powder/routines/absorb_corrections.py:
```diff
@@ -112,3 +112,29 @@ def _do_cylinder_absorb_corrections(ws_to_correct, multiple_scattering, is_vanad
+                         ContainerMaterial=container_material)
+    else:
+        mantid.SetSample(workspace, Geometry=sample_details.generate_sample_geometry(),
+                         Material=sample_details.generate_sample_material())
```
I'm not sure this code path works - if a container isn't specified (eg if I comment out the sample_details.set_container line in the test script) I get an error from the ApplyPaalmanPingsCorrection line:
`Error in execution of algorithm ApplyPaalmanPingsCorrection:
'acc'
  at line 89 in 'C:/mantid/Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ApplyPaalmanPingsCorrection.py'
`
I think if the container isn't set then the CanWorkspace parameter shouldn't be supplied lower down on L138

This only happens if someone specifies empty_can_subtraction_method=PaalmanPings but then doesn't set up a container. But worth fixing up I think

DannyHindson reviewed (changes requested):
Last round of changes look good thanks. I've found two or three small things on different topics. Hopefully quick!

DanielMurphy22 committed:
Allow containerless path for paalmanings

Also bolster docs comment
Improve error message for sample_details FlatPlateholder

DannyHindson commented on scripts/Diffraction/isis_powder/routines/sample_details.py:
```diff
@@ -146,6 +195,42 @@ def thickness(self):
+    def generate_container_material(self):
+        if self.container_material_object:
+            return {'ChemicalFormula': self.container_material_object.chemical_formula}
+        else:
```
If you run a reduction where the container material is a compound (not v realistic but say H2-O for example) you get an error from the SetSample algorithm:
```
The number density could not be determined. Please provide the number density, ZParameter and unit cell volume or mass density.
```
My sample_details object was set up like this:
```
sample_details = SampleDetails(width=4.0, height=4.0, thickness=0.1, center=[0, 0, 0], shape='slab', angle=0.0)
sample_details.set_material(chemical_formula='Si', number_density=2.0, crystal_density=1.0)
sample_details.set_container(front_thick=0.3175, back_thick=0.3175, chemical_formula='H2-O',number_density=1.0)
polaris.set_sample_details(sample=sample_details)
```


Don't think Mantid can look up the number density for compounds in the same way it can if the material is just an element.
I think you need to copy in something similar to lines 214-219 from generate_sample_material in here

DanielMurphy22 committed:
Allow compounds for container material

DannyHindson commented on scripts/Diffraction/isis_powder/routines/sample_details.py:
```diff
@@ -146,6 +195,46 @@ def thickness(self):
+        else:
+            return None
+
+    def generate_container_material(self):
```
You still need this if else statement around the rest of the lines in case no container is set:
```
if self.container_material_object:
...
else:
   return None
```
This prevents an error if you do a PP run without having called sample_details.set_container

DannyHindson reviewed (approved):
All looks good now! Good job

Since we've changed the code in focus.py I think we should get someone from the instruments that use these scripts to do some beta testing before the next release.

It's possible at some point the users may want some more parameters on the PaalmanPingsMonteCarloAbsorption algorithm exposed in the .yaml file (eg SparseInstrument) but let's get this out there and see what Helen thinks
